src/blocks/privacy/pipewire.rs

```diff
@@ -97,75 +100,65 @@ impl Client {
-                }
-            })
-            .register();
+        let updated = Rc::new(Cell::new(false));
+        let updated_copy = updated.clone();
+        let updated_copy2 = updated.clone();
 
         // Register a callback to the `global` event on the registry, which notifies of any new global objects
         // appearing on the remote.
         // The callback will only get called as long as we keep the returned listener alive.
         let _registry_listener = registry
             .add_listener_local()
             .global(move |global| {
-                let global_id = global.id;
                 let Some(global_props) = global.props else {
                     return;
                 };
-                if global.type_ == ObjectType::Node {
-                    client
-                        .data
-                        .lock()
-                        .unwrap()
-                        .nodes
-                        .insert(global_id, Node::new(global_id, global_props));
-                    client.send_update_event();
-                } else if global.type_ == ObjectType::Link {
-                    let Some(link) = Link::new(global_props) else {
-                        return;
-                    };
-                    client.data.lock().unwrap().links.insert(global_id, link);
-                    client.send_update_event();
+                match &global.type_ {
+                    ObjectType::Node => {
+                        client
+                            .data
+                            .lock()
+                            .unwrap()
+                            .nodes
+                            .insert(global.id, Node::new(global.id, global_props));
+                        updated_copy.set(true);
+                    }
+                    ObjectType::Link => {
+                        let Some(link) = Link::new(global_props) else {
+                            return;
+                        };
+                        client.data.lock().unwrap().links.insert(global.id, link);
+                        updated_copy.set(true);
+                    }
+                    _ => (),
                 }
             })
             .global_remove(move |uid| {
                 let mut data = client.data.lock().unwrap();
-                if data.nodes.remove(&uid).is_some() {
-                    client.send_update_event();
-                }
-                if data.links.remove(&uid).is_some() {
-                    client.send_update_event();
+                if data.nodes.remove(&uid).is_some() || data.links.remove(&uid).is_some() {
+                    updated_copy2.set(true);
                 }
             })
             .register();
 
-        main_loop.run();
-    }
-
-    fn add_event_listener(&self, tx: UnboundedSender<()>) {
-        self.event_listeners.lock().unwrap().push(tx);
+        loop {
+            main_loop.loop_().iterate(Duration::from_secs(60 * 60 * 24));
```

Comment:
Why duration of a day instead of no timeout?
https://pipewire.pages.freedesktop.org/pipewire-rs/pipewire/loop_/struct.Loop.html: 

pub fn [iterate](https://pipewire.pages.freedesktop.org/pipewire-rs/pipewire/loop_/struct.Loop.html#method.iterate)(&self, timeout: [Duration](https://doc.rust-lang.org/1.71.1/core/time/struct.Duration.html)) -> [i32](https://doc.rust-lang.org/1.71.1/std/primitive.i32.html)
Perform one iteration of the loop.

An optional timeout can be provided. 0 for no timeout, -1 for infinite timeout.

This function will block up to the provided timeout and then dispatch the fds with activity. The number of dispatched fds is returned.

Reply:
https://gitlab.freedesktop.org/pipewire/pipewire-rs/-/issues/100

Reply:
My understanding is that setting 0 for the timeout is the optional timeout. I imagine it's done like this since there's no `Option` in C so it just takes an int.

Reply:
No timeout means "nonblocking", but we need "infinitely blocking", which is achieved in C by passing -1. `Duration` cannot represent negative time intervals.